### PR TITLE
WIP: new-test

### PR DESCRIPTION
**Pull request by phadej**
I have no idea what I have done, so all comments are more than welcome.

### CODE REVIEW

File: cabal-install/Distribution/Client/ProjectPlanning.hs [hunk 1]
```diff
@@ -1877,13 +1887,17 @@ elaboratePackageTargets ElaboratedConfiguredPackage{..} targets =
                       $ targets
         buildHaddocks = HaddockDefaultComponents `elem` targets
 
-     in (buildTargets, replTargets, buildHaddocks)
+     in (buildTargets, testTargets, replTargets, buildHaddocks)
   where
     --TODO: need to report an error here if defaultComponents is empty
     elaborateBuildTarget  BuildDefaultComponents    = pkgDefaultComponents
     elaborateBuildTarget (BuildSpecificComponent t) = [t]
```
Comment (phadej) [inline cabal-install/Distribution/Client/ProjectPlanning.hs]: I guess I have to include test targets here as well.

**Comment by ezyang** [comment]
Big question from me is how this relates to #4179.

**Comment by phadej** [comment]
I wasnt' aware of #4179. And it's looks like bitrotten, and too big, I don't like refactors and new features in single diff. FWIW, I don't see #4179 going in soon enough. This on other hand is ~200 lines diff.

**Comment by 23Skidoo** [comment]
I'd like @dcoutts to weigh in before merging.

**Comment by phadej** [comment]
FWIW, I use this with `servant` repository. The output is a bit misleading, but e.g .`cabal new-test --enable tests servant` does there exactly what I want.

**Comment by phadej** [comment]
... i.e. I'm totally OK with @dcoutts eventually refactoring ProjectPlanning etc, but this is valuable already today.

File: cabal-install/Distribution/Client/CmdTest.hs [hunk 2]
```diff
@@ -0,0 +1,88 @@
+{-# LANGUAGE NamedFieldPuns #-}
+
+-- | cabal-install CLI command: repl
```
Comment (ezyang) [inline cabal-install/Distribution/Client/CmdTest.hs]: typo

**Comment by ezyang** [comment]
Irrelevant of anything, this needs tests. In this case, I would advise just finding the tests in cabal-testsuite which exercise Setup test, cabal.project'izing them and then running `cabal new-test` to run them. Let me know if you need more guidance on this.

**Comment by phadej** [comment]
@ezyang does the last commit looks somewhat right?

**Comment by ezyang** [comment]
Looks good, but put it in a separate file (current convention is `cabal.test.hs`) so that it can be run in parallel with the setup variant.

**Comment by phadej** [comment]
I'll wait till travis says it's otherwise ok.

**Comment by phadej** [comment]
Separated the tests, travis is green.

**Comment by ezyang** [comment]
Is this non WIP now?

**Comment by phadej** [comment]
Well, it's not "ready, but it's usable.

File: cabal-install/Distribution/Client/CmdTest.hs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  commandUsage        = usageAlternatives "new-test" [ "[FLAGS] TARGET" ],
+  commandDescription  = Just $ \_ -> wrapText $
+        "Build and run test targets",
+  commandNotes        = Just $ \_pname ->
+        "Examples:\n"
+   }
+
+-- | The @test@ command is very much like @build@. It brings the install plan
+-- up to date, selects that part of the plan needed by the given or implicit
+-- test arget(s) and then executes the plan.
```
Comment (ezyang) [inline cabal-install/Distribution/Client/CmdTest.hs]: arget?

**Comment by phadej** [comment]
I'd rather add `new-test` as hidden command. Unfortunately I don't have time to polish it in near future (next month or so), especially if @dcoutts refactor will redo it anyway.

**Comment by ezyang** [comment]
That's OK by me. At least, reading the patch, I don't think the *semantics* are wrong, so even if it's buggy it may still be fine.

It's possible that @dcoutts' refactor won't actually affect this, because actually 'cabal new-run' works quite differently form 'cabal new-test', and from reading this test it does seem like new-test will work fine in this realm.

**Pull request merged**